File: daily/1021. 删除最外层的括号.js
/**
 * @param {string} S
 * @return {string}
 */
var removeOuterParentheses = function(S) {
    var ch=[],res='';   //ch为栈，res存放返回结果
    for(let i=0;i<S.length;i++){
        if(S[i]=='('){
            if(ch.length>0) //最外层的不被加入res中
                res+=S[i];
            ch.push(S[i]);
        }
        else{
            res+=S[i];
            ch.pop();
            if(ch.length==0)    //最外层的不被加入res中
                res=res.substr(0,res.length-1);
        }
    }
    return res;
};

File: daily/C++版本/1021. 删除最外层的括号.cpp
class Solution {
public:
    string removeOuterParentheses(string S) {
        string res;
        stack<char> stack;
        for(int i=0;i<S.length();i++){
            if(S[i]=='('){
                if(!stack.empty())
                    res+=S[i];
                stack.push(S[i]);
            }
            else{
                res+=S[i];
                stack.pop();
                if(stack.empty()) 
                    res.pop_back();          
            }
        }
        return res;
    }
};
